src/SIL.LCModel.Core/Text/TsStrBldr.cs

```diff
@@ -319,5 +319,59 @@ public void Clear()
 			m_runs.Clear();
 			m_runs.Add(TsRun.EmptyRun);
 		}
+
+		private class CachedStringBuilder
+		{
+			private string _cachedString;
+			private StringBuilder _stringBuilder;
+
+			public CachedStringBuilder(string initialContent)
+			{
+				_cachedString = initialContent;
+				_stringBuilder = new StringBuilder();
+			}
+
+			public string StringValue
+			{
+				get
+				{
+					if (_cachedString == null)
+					{
+						_cachedString = _stringBuilder.ToString();
+						_stringBuilder.Clear();
+					}
+
+					return _cachedString.Length == 0 ? null : _cachedString;
+				}
+			}
+
+			public void Clear()
+			{
+				_cachedString = null;
+				_stringBuilder.Clear();
+			}
+
+			public void Remove(int ichMin, int i)
+			{
+				if (_cachedString != null)
+				{
+					_stringBuilder.Append(_cachedString);
+					_cachedString = null;
+				}
+
+				_stringBuilder.Remove(ichMin, i);
+			}
+
+			public void Insert(int ichMin, string insertText)
+			{
```

Comment:
You could add an assert like `Debug.Assert(_cachedString == null || _stringBuilder.Length == 0);`